### PR TITLE
Rewritten tab without jquery

### PR DESCRIPTION
**Pull request by alekitto**
Related to #23586

### CODE REVIEW

File: js/src/dom/selectorEngine.js [hunk 1]
```diff
@@ -50,12 +48,71 @@ const SelectorEngine = (() => {
...
+  } else {
+    findFn = function (selector) {
+      if (!scopeSelectorRegex.test(selector)) {
+        return this.querySelectorAll(selector)
+      }
+
+      const hasId = Boolean(this.id)
+      if (!hasId) {
+        // eslint-disable-next-line no-magic-numbers
+        this.id = `q${Math.floor(Math.random() * 9000000) + 1000000}`
```
Comment (Johann-S) [inline js/src/dom/selectorEngine.js]: Use `Util.getUID` (https://github.com/twbs/bootstrap/blob/v4-without-jquery/js/src/util.js#L51) instead of that

File: js/src/tab.js [hunk 2]
```diff
@@ -46,10 +46,10 @@ const Tab = (() => {
     DROPDOWN              : '.dropdown',
     NAV_LIST_GROUP        : '.nav, .list-group',
     ACTIVE                : '.active',
-    ACTIVE_UL             : '> li > .active',
+    ACTIVE_UL             : ':scope > li > .active',
```
Comment (Johann-S) [inline js/src/tab.js]: why did you changed that selector ?

Comment (alekitto) [reply js/src/tab.js]: `> li > .active` is not a valid selector. Using with `querySelector/querySelectorAll` throws a `SyntaxError`.

Comment (Johann-S) [reply js/src/tab.js]: yep so you should find `li` children and after that `.active` elements, no need here to implement `:scope`

Comment (alekitto) [reply js/src/tab.js]: Sorry, can't understand. Are you suggesting not to use a selector? Or a double use of `.children`? In both cases it could be very slow

File: js/src/dom/selectorEngine.js [hunk 3]
```diff
@@ -50,12 +48,70 @@ const SelectorEngine = (() => {
...
+          this.removeAttribute('id')
+        }
+      }
+
+      return nodeList
+    }
+  }
+
+  let findOneFn = null
+  if (supportScopeQuery) {
```
Comment (Johann-S) [inline js/src/dom/selectorEngine.js]: you should merge those two `if`

File: js/src/dom/selectorEngine.js [hunk 4]
```diff
@@ -29,17 +31,13 @@ const SelectorEngine = (() => {
   if (!Element.prototype.closest) {
     fnClosest = (element, selector) => {
       let ancestor = element
-      if (!document.documentElement.contains(element)) {
```
Comment (Johann-S) [inline js/src/dom/selectorEngine.js]: this polyfill come from MDN, why did you change that ?

File: js/src/tab.js (hunk 2 above)
Comment (Johann-S) [reply js/src/tab.js]: Nope you're right that's better like that, sorry I didn't know `:scope` before your PR :smile:

File: js/src/dom/selectorEngine.js [hunk 5]
```diff
@@ -64,21 +120,24 @@ const SelectorEngine = (() => {
...
+      return findFn.call(element, selector)
     },
 
-    findOne(selector, element = document) {
+    findOne(selector, element = document.documentElement) {
       if (typeof selector !== 'string') {
         return null
       }
 
-      let selectorType = 'querySelector'
```
Comment (Johann-S) [inline js/src/dom/selectorEngine.js]: Not a big fan of that, because if it's an ID you'll use `querySelectorAll` instead of `getElementById` which is faster

**Review by Johann-S — CHANGES_REQUESTED** [review]

File: js/src/dom/selectorEngine.js (hunk 3 above)
Comment (alekitto) [reply js/src/dom/selectorEngine.js]: Ok

File: js/src/dom/selectorEngine.js (hunk 4 above)
Comment (alekitto) [reply js/src/dom/selectorEngine.js]: Tests are failing with that `if`. In any case, the DOM Standard specification (https://dom.spec.whatwg.org/#dom-element-closest) doesn't require the element to be a discendant of the `document` object

Comment (Johann-S) [reply js/src/dom/selectorEngine.js]: that's right but I don't see a use case where the element should be outside of the `document.documentElement` :thinking:

Comment (alekitto) [reply js/src/dom/selectorEngine.js]: Probably some tests don't add elements to the DOM before execution. If you prefer, I can revert this change and fix those tests.

Comment (Johann-S) [reply js/src/dom/selectorEngine.js]: No that's ok :+1:

File: js/src/dom/selectorEngine.js (hunk 5 above)
Comment (alekitto) [reply js/src/dom/selectorEngine.js]: Multiple selector starting with `#` (ex: `#id > li, #id1 > .active`) will match and return a wrong result

Comment (alekitto) [reply js/src/dom/selectorEngine.js]: However, I think this is in the wrong PR, i've encountered an error on this while working on ScrollSpy 😅

Comment (alekitto) [reply js/src/dom/selectorEngine.js]: @Johann-S Should I revert this? Or fix testing the selector against a regex that excludes complex selectors?

File: js/src/tab.js (hunk 2 above)
Comment (artemsky) [reply js/src/tab.js]: :scope deprecated?) https://developer.mozilla.org/en-US/docs/Web/CSS/:scope

Comment (alekitto) [reply js/src/tab.js]: Nothing indicates that `:scope` selector has been deprecated. It is part of the CSS Selectors Level 4 Working Draft. Some browsers have removed  the `scoped` style tag attribute, but the `:scope` pseudo-class is still supported and works correctly in CSS (works the same as `:root`) and in `querySelector/querySelectorAll`

File: js/src/dom/selectorEngine.js (hunk 5 above)
Comment (Johann-S) [reply js/src/dom/selectorEngine.js]: Nope finally I think it's fine like this 👍

**Comment by Johann-S** [comment]
@alekitto can you update your branch please ? 👍

**Comment by alekitto** [comment]
@Johann-S done 👍

File: js/src/tab.js [hunk 6]
```diff
@@ -171,54 +164,49 @@ const Tab = (($) => {
...
       }
 
-      $(element).addClass(ClassName.ACTIVE)
+      element.classList.add(ClassName.ACTIVE)
       if (element.getAttribute('role') === 'tab') {
         element.setAttribute('aria-selected', true)
       }
 
-      if (isTransitioning) {
+      if (isTransitioning || !active) {
```
Comment (Johann-S) [inline js/src/tab.js]: It seems an unrelated change here 🤔 Maybe can you explain it ?

Comment (alekitto) [reply js/src/tab.js]: This check was introduced to fix fading when no initial panel was active, but i've probably missed a test. I'll push it asap, but if you want to keep things separated, i can revert this check and submit a pr just for this case.

Comment (Johann-S) [reply js/src/tab.js]: Is it an issue in our current `v4-dev` branch ?

Comment (alekitto) [reply js/src/tab.js]: I’ve just tested it on the v4-dev, the issue is present in the visual test page

Comment (Johann-S) [reply js/src/tab.js]: So I prefer if you can create a PR (which target `v4-dev`) to resolve that or if you can create an issue which describe that

Comment (alekitto) [reply js/src/tab.js]: Correct. I've pushed a PR to explain and resolve the issue on `v4-dev` branch

Comment (alekitto) [reply js/src/tab.js]: I've reverted this change for this PR, as it is not related

**Review by Johann-S — APPROVED** [review]
